presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -315,15 +317,32 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
             }
         }
 
+        leftPlanBuilder = subqueryPlanner.handleSubqueries(leftPlanBuilder, preLeftConditions, node);
+        leftPlanBuilder = wrapWithFilterIfNeeded(leftPlanBuilder, preLeftConditions);
+        rightPlanBuilder = subqueryPlanner.handleSubqueries(rightPlanBuilder, preRightConditions, node);
+        rightPlanBuilder = wrapWithFilterIfNeeded(rightPlanBuilder, preRightConditions);
+
         PlanNode root = new JoinNode(idAllocator.getNextId(),
                 JoinNode.Type.typeConvert(node.getType()),
-                wrapWithFilterIfNeeded(leftPlanBuilder.getRoot(), preLeftConditions),
-                wrapWithFilterIfNeeded(rightPlanBuilder.getRoot(), preRightConditions),
+                leftPlanBuilder.getRoot(),
+                rightPlanBuilder.getRoot(),
                 equiClauses.build(),
                 Optional.empty(),
                 Optional.empty(),
                 Optional.empty());
 
+        if (node.getType() != INNER) {
+            List<InPredicate> inPredicateSubqueries = analysis.getInPredicateSubqueries(node);
```

Comment:
test cases when complex expression is "complex" but touches just one side of the join.
Test both outer and inner side.